### ISSUE TITLE
NonZeroUuid

### ISSUE THREAD

**Issue by rick-de-water** [uuid-rs/uuid#770]
Is there a `NonZeroUuid` type, or has one been considered, similar to the other [`NonZero`](https://doc.rust-lang.org/std/num/struct.NonZero.html) types, such that the following is true?

```rust
assert_eq!(std::mem::size_of::<Option<NonZeroUuid>>(), std::mem::size_of::<Uuid>());
```

**Comment by KodrAus** [comment]
Hi @rick-de-water 👋 A UUID with all zeroes is actually a valid value, it’s called the [nil UUID](https://www.rfc-editor.org/rfc/rfc9562.html#name-nil-uuid).

Are you looking at `Option<Uuid>` to avoid serializing them if they’re nil?

**Comment by rick-de-water** [comment]
Of course, and an integer with a value of zero is also valid. There are however situation where you know that you will never use that value, and you can allow the compiler to do niche optimization if you inform it. This is what happens with the [`NonZero`](https://doc.rust-lang.org/std/num/struct.NonZero.html) types, and of course with references, since they also cannot be null.

Its actually quite common to have domain models that do not have NIL UUIDs, since they get generated at the moment of creation and never change. Having `NonZeroUuid`s (or `NonNilUuid`s if you want to be specific) can have a definite impact on the memory usage of a program.

**Comment by KodrAus** [comment]
Coming back to this, it does seem like something that would be reasonable to add support for. I like the idea of calling it `NonNilUuid`. To avoid duplicating the entire API surface area though, I think it should be limited to conversions to and from `Uuid`s, and trait implementations that are also on `Uuid`.
